scrapy_autoextract/providers.py

```diff
@@ -0,0 +1,98 @@
+from typing import ClassVar, Optional, Type
+
+from autoextract.aio import request_raw
+from autoextract.request import Request as AutoExtractRequest
+from autoextract_poet.page_inputs import (
+    AutoExtractArticleData,
+    AutoExtractProductData,
+)
+from scrapy import Request
+from scrapy.settings import Settings
+from scrapy.statscollectors import StatsCollector
+from scrapy_poet.page_input_providers import (
+    PageObjectInputProvider,
+    register,
+)
+
+
+class _Provider(PageObjectInputProvider):
+    """An interface that describes a generic AutoExtract Provider.
+
+    It should not be used publicly as it serves the purpose of being a base
+    class for more specific providers such as Article and Product providers.
+    """
+
+    provided_class: ClassVar[Optional[Type]]
+
+    def __init__(
+            self,
+            request: Request,
+            settings: Settings,
+            stats: StatsCollector,
+    ):
+        """Initialize provider storing its dependencies as attributes."""
+        self.request = request
+        self.stats = stats
+        self.settings = settings
+
+    async def __call__(self):
+        """Make an AutoExtract request and build a Page Input of provided class
+        based on API response data.
+        """
+        page_type = self.get_page_type()
+        self.stats.inc_value(f"autoextract/{page_type}/total")
+
+        request = AutoExtractRequest(
+            url=self.request.url,
+            pageType=page_type,
+        )
+        api_key = self.settings.get("AUTOEXTRACT_USER")
```

Comment:
Here we force to define that in the settings, but it would be good that by default it can be picked up from `SCRAPINGHUB_AUTOEXTRACT_KEY` environment variable, as is done in https://github.com/scrapinghub/scrapinghub-autoextract. 

That way we would have a way of configuring the key a single time and use it from all the Scrapinghub tools.

In other words, if `AUTOEXTRACT_USER` is not in the settings, then just rely on the ENV variable. 

Also probably I would change the name to `AUTOEXTRACT_KEY` to be coherent.

Reply:
- The reason we're naming it `AUTOEXTRACT_USER` instead of `AUTOEXTRACT_KEY` is to keep it consistent with the middleware (see [here](https://github.com/scrapinghub/scrapy-autoextract/blob/master/scrapy_autoextract/middlewares.py#L50)).
- The `api_key` argument is an optional string that fallbacks to the environment variable  `SCRAPINGHUB_AUTOEXTRACT_KEY` if it's `None` (see [here](https://github.com/scrapinghub/scrapinghub-autoextract/blob/9e7f3cc03208d54b858708b8f556acf7d312335b/autoextract/apikey.py#L12)).

Reply:
You could propose the change from `AUTOEXTRACT_USER` to `AUTOEXTRACT_KEY` in a new issue, what do you think, @ivanprado?

Reply:
+1 to discuss rename as a separate issue. Regarding ENV variable, if this works, it'd be good to document that, and check (at least manually).

Reply:
Agree.

Reply:
Created an issue here: https://github.com/scrapinghub/scrapy-autoextract/issues/14

Reply:
> Regarding ENV variable, if this works, it'd be good to document that, and check (at least manually).

It works because my sample spider does not define `AUTOEXTRACT_USER` on `settings.py`.
I've updated the readme file describing this setting as optional when using providers and mentioning `SCRAPINGHUB_AUTOEXTRACT_KEY` environment variable as its default value.